server/src/main/java/org/opensearch/common/cache/tier/TieredSpilloverCache.java

```diff
@@ -0,0 +1,292 @@
+    private final Optional<StoreAwareCache<K, V>> onDiskCache;
+    private final StoreAwareCache<K, V> onHeapCache;
+    private final StoreAwareCacheEventListener<K, V> listener;
+    ReadWriteLock readWriteLock = new ReentrantReadWriteLock();
+    ReleasableLock readLock = new ReleasableLock(readWriteLock.readLock());
+    ReleasableLock writeLock = new ReleasableLock(readWriteLock.writeLock());
+
+    /**
+     * Maintains caching tiers in ascending order of cache latency.
+     */
+    private final List<StoreAwareCache<K, V>> cacheList;
+
+    TieredSpilloverCache(Builder<K, V> builder) {
+        Objects.requireNonNull(builder.onHeapCacheBuilder, "onHeap cache builder can't be null");
+        this.onHeapCache = builder.onHeapCacheBuilder.setEventListener(this).build();
+        if (builder.onDiskCacheBuilder != null) {
+            this.onDiskCache = Optional.of(builder.onDiskCacheBuilder.setEventListener(this).build());
+        } else {
+            this.onDiskCache = Optional.empty();
+        }
+        this.listener = builder.listener;
+        this.cacheList = this.onDiskCache.map(diskTier -> Arrays.asList(this.onHeapCache, diskTier)).orElse(List.of(this.onHeapCache));
+    }
+
+    // Package private for testing
+    StoreAwareCache<K, V> getOnHeapCache() {
+        return onHeapCache;
+    }
+
+    // Package private for testing
+    Optional<StoreAwareCache<K, V>> getOnDiskCache() {
+        return onDiskCache;
+    }
+
+    @Override
+    public V get(K key) {
+        StoreAwareCacheValue<V> cacheValue = getValueFromTieredCache(true).apply(key);
+        if (cacheValue == null) {
+            return null;
+        }
+        return cacheValue.getValue();
+    }
+
+    @Override
+    public void put(K key, V value) {
+        try (ReleasableLock ignore = writeLock.acquire()) {
+            onHeapCache.put(key, value);
+            listener.onCached(key, value, CacheStoreType.ON_HEAP);
+        }
+    }
+
+    @Override
+    public V computeIfAbsent(K key, LoadAwareCacheLoader<K, V> loader) throws Exception {
+        // We are skipping calling event listeners at this step as we do another get inside below computeIfAbsent.
+        // Where we might end up calling onMiss twice for a key not present in onHeap cache.
+        // Similary we might end up calling both onMiss and onHit for a key, in case we are received concurrent
+        // requests for the same key which requires loading only once.
+        StoreAwareCacheValue<V> cacheValue = getValueFromTieredCache(false).apply(key);
+        if (cacheValue == null) {
+            // Add the value to the onHeap cache. We are calling computeIfAbsent which does another get inside.
+            // This is needed as there can be many requests for the same key at the same time and we only want to load
+            // the value once.
+            V value = null;
+            try (ReleasableLock ignore = writeLock.acquire()) {
```

Comment:
Yeah ideally onHeapCache should take care. I added locks here to avoid same key going to both onheap/disk cache in a rare scenario.
What about this scenario:
1. We do a get on key2 which is not present on tiered cache. ```onHeapCache.computeIfAbsent``` gets called here. Suppose key1 gets evicted(in case onHeapCache was at its capacity) in the same thread. Eviction flow is underway where key1 is going to put under disk cache.
2. We do a parallel get on key1 on different thread, get() returns null as it neither on onHeap nor disk at this moment. Here ```onHeapCache.computeIfAbsent``` will get called and succeed, meanwhile the disk put from step#1 also will succeed.
Resulting same key present on both caches.

Let me know if this make sense.

Reply:
Makes sense. When `computeIfAbsent` gets called from multiple threads that lead to this.